custom-metrics-stackdriver-adapter/pkg/adapter/translator/utils/filter_builder.go

```diff
@@ -0,0 +1,144 @@
+package utils
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+
+	"k8s.io/klog"
+)
+
+type Schema struct {
+	resourceType string
+	metricType   string
+	project      string
+	cluster      string
+	location     string
+	namespace    string
+	pods         string
+}
+
+const (
+	PodType        = "k8s_pod"
+	PrometheusType = "prometheus_target"
+	LegacyType     = "<not_allowed>"
+)
+
+var (
+	PodSchema = &Schema{
+		resourceType: "resource.type",
+		metricType:   "metric.type",
+		project:      "resource.labels.project_id",
+		cluster:      "resource.labels.cluster_name",
+		location:     "resource.labels.location",
+		namespace:    "resource.labels.namespace_name",
+		pods:         "resource.labels.pod_name",
+	}
+	LegacyPodSchema = &Schema{
+		resourceType: "",
+		metricType:   "metric.type",
+		project:      "resource.labels.project_id",
+		cluster:      "resource.labels.cluster_name",
+		location:     "resource.labels.location",
+		namespace:    "resource.labels.namespace_name",
+		pods:         "resource.labels.pod_id",
+	}
+	PrometheusSchema = &Schema{
+		resourceType: "resource.type",
+		metricType:   "metric.type",
+		project:      "resource.labels.project_id",
+		cluster:      "resource.labels.cluster",
+		location:     "resource.labels.location",
+		namespace:    "resource.labels.namespace",
+		pods:         "metric.labels.pod",
+	}
+	SchemaTypes = map[string]string{
+		"pod":        PodType,
+		"prometheus": PrometheusType,
+		"legacy":     LegacyType,
+	}
+)
+
+type FilterBuilder struct {
+	schema  *Schema
+	filters []string
+}
+
+// initialize with resource type and whether to use legacy model
+// For legacy pod, use NewFilterBuilder(AnyType, false)
+func NewFilterBuilder(resourceType string) *FilterBuilder {
+	var schema *Schema
+	switch resourceType {
+	case PrometheusType:
+		schema = PrometheusSchema
+	case LegacyType:
+		schema = LegacyPodSchema
+	default:
```

Comment:
Can you make the PodSchema a dedicated too so it's clear that it's expected?
The default can stay as is, I'd just prefer all defined cases to be specified as such.

Reply:
sounds good to me